### PR TITLE
Add Support for Panorama/PAN-OS Master Key Rotation

### PR DESCRIPTION
**Pull request by kamalq97**
## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-11635)

## Description
- Added new commands to manage the master key across PAN-OS devices and Panorama:
  - ***pan-os-create-master-key*** - ‼️ potentially harmful command
  - ***pan-os-update-master-key*** - ‼️ potentially harmful command
  - ***pan-os-get-master-key-details***

> [!CAUTION]
>  Creating or updating the encryption master key of Palo Alto Networks Firewall or Panorama invalidates the current API key and requires obtaining a new one. All subsequent commands will raise an "Invalid Credentials" error until a new API key is obtained and the integration instance is updated accordingly.

### CODE REVIEW

**Comment by kamalq97** [comment]
There is a failing test playbook `test_playbook_pan_os_firewall_exceptions_commands` that is being handled in the Dev Bug ticket [CIAC-12199](https://jira-dc.paloaltonetworks.com/browse/CIAC-12199)

**Comment by kamalq97** [comment]
@dorschw @ShirleyDenkberg Please advise on where and how to communicate the important limitation / warning mentioned in the pull request description above ☝️

**Comment by ShirleyDenkberg** [comment]
@dorschw @jlevypaloalto Doc review completed.

File: Packs/PAN-OS/Integrations/Panorama/Panorama_description.md [hunk 1]
```diff
@@ -1,6 +1,10 @@
 The integration uses the Panorama XML API.
-To obtain an API Key, run the following REST command and copy the key:
-https://[PanoramaIP]/api/?type=keygen&user=[user]&password=[password]
+To obtain an API key, run the following CURL command and copy the key:
+```shell
+-curl -H "Content-Type: application/x-www-form-urlencoded" -X POST https://[PanoramaIP]/api/\?type\=keygen -d 'user=[user]&password=[password]'
+```
+
+**Creating or updating the encryption master key of Palo Alto Networks Firewall or Panorama invalidates the current API key and requires obtaining a new one.**
```
Comment (ShirleyDenkberg) [inline Packs/PAN-OS/Integrations/Panorama/Panorama_description.md:7]: Creating or updating the encryption master key of Palo Alto Networks Firewall or Panorama invalidates the current API key and requires obtaining a new one. All subsequent commands will raise an "Invalid Credentials" error until a new API key is obtained and the integration instance is updated accordingly.

File: Packs/PAN-OS/Integrations/Panorama/README.md [hunk 2]
```diff
@@ -48,7 +48,11 @@ This integration was integrated and tested with versions 8.xx, 9.xx, and 10.xx o
    * [pan-os-query-logs](#pan-os-query-logs)
    * [pan-os-check-logs-status](#pan-os-check-logs-status)
    * [pan-os-get-logs](#pan-os-get-logs)
+   * [pan-os-get-master-key-details](#pan-os-get-master-key-details)
+   * [pan-os-create-master-key](#pan-os-create-master-key)
+   * [pan-os-update-master-key](#pan-os-update-master-key)
 * The target argument is supported only in operational type commands. Meaning, you cannot use it with commit, logs, or PCAP commands.
+* After running the `pan-os-create-master-key` or `pan-os-update-master-key` command, the current API key will become invalid and a new one needs to be obtained. Refer to the [documentation on how to get an API key](https://docs.paloaltonetworks.com/pan-os/11-0/pan-os-panorama-api/get-started-with-the-pan-os-xml-api/get-your-api-key).
```
Comment (ShirleyDenkberg) [inline Packs/PAN-OS/Integrations/Panorama/README.md:56]:
```suggestion

Creating or updating the encryption master key of Palo Alto Networks Firewall or Panorama invalidates the current API key and requires obtaining a new one. All subsequent commands will raise an "Invalid Credentials" error until a new API key is obtained and the integration instance is updated accordingly.

```

File: Packs/PAN-OS/Integrations/Panorama/Panorama_description.md (hunk 1 above)
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama_description.md]: Thank you for your feedback. The sentence has been updated based on your suggestion in commit 85dab2f2c6d1f0308f5d274059664b920a3ce75a.

File: Packs/PAN-OS/Integrations/Panorama/README.md (hunk 2 above)
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/README.md:56]: Thank you for your feedback. The sentence has been updated based on your suggestion in commit 85dab2f2c6d1f0308f5d274059664b920a3ce75a.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 3]
```diff
@@ -14511,6 +14511,116 @@ def pan_os_list_profile_exception_command(args: dict) -> CommandResults:
...
+    """
+    master_key_cmd = build_master_key_create_or_update_cmd(args, action=action)
+    raw_response: dict = http_request(URL, 'GET', params={'type': 'op', 'key': API_KEY, 'cmd': master_key_cmd})
+    response_result = dict_safe_get(raw_response, ('response', 'result'))  # human readable message
+
+    # Creating or updating the encryption master key by definition invalidates the current API key, refer to the integration docs.
+    demisto.info(f'The master key of {URL} has been {action}d. The current API key has been invalidated.')
+
+    return CommandResults(
+        readable_output=f'{response_result}. The current API key has been invalidated. Generate a new API key and ensure the integration instance is updated accordingly.',
```
Comment (dorschw) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:14564]:
```suggestion
        readable_output=f'{response_result}. ⚠️ The current API key is no longer valid! (by design) Generate a new API key and update it in the integration instance configuration to keep using the integration.',
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (dorschw) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:1]: in the `description` of any command that invalidates the API key, append a short message saying it.

File: Packs/PAN-OS/Integrations/Panorama/README.md
Comment (dorschw) [inline Packs/PAN-OS/Integrations/Panorama/README.md:1]: generate the docs after applying the yml comments, so the README matches it.

**Review by dorschw — COMMENTED** [review]
Well done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 4]
```diff
@@ -14511,6 +14511,116 @@ def pan_os_list_profile_exception_command(args: dict) -> CommandResults:
...
+    """
+    master_key_args = [
+        add_argument(arg=args.get('lifetime_in_hours'), field_name='lifetime', member=False),
+        add_argument(arg=args.get('reminder_in_hours'), field_name='reminder', member=False),
+    ]
+    if action == 'create':
+        master_key_args.append(
+            add_argument(arg=args.get('master_key'), field_name='new-master-key', member=False),
+        )
+    else:
```
[the remark this answers is not in the document]
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Added pattern matching with a `ValueError` being raised under `case _` in commit 77379b70697a17549d3313dd6b3d99ddcb0cc026

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 5]
```diff
@@ -14511,6 +14511,116 @@ def pan_os_list_profile_exception_command(args: dict) -> CommandResults:
...
+    """Builds the XML command for creating or updating the default master key on Panorama / PAN-OS.
+
+    Args:
+        args (dict): The command arguments.
+        action ('create'| 'update'): Whether to create a new master key or update an existing one.
+
+    Returns:
+        str: XML string of the master key create or update command.
+    """
+    master_key_args = [
```
[the remark this answers is not in the document]
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Renamed to `xml_args` in commit 77379b70697a17549d3313dd6b3d99ddcb0cc026

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 6]
```diff
@@ -14511,6 +14511,116 @@ def pan_os_list_profile_exception_command(args: dict) -> CommandResults:
...
+    Args:
+        args (dict): The command arguments.
+        action ('create'| 'update'): Whether to create a new master key or update an existing one.
+
+    Returns:
+        CommandResults: Contains readable output and raw response.
+    """
+    master_key_cmd = build_master_key_create_or_update_cmd(args, action=action)
+    raw_response: dict = http_request(URL, 'GET', params={'type': 'op', 'key': API_KEY, 'cmd': master_key_cmd})
+    response_result = dict_safe_get(raw_response, ('response', 'result'))  # human readable message
```
[the remark this answers is not in the document]
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: You are right, `http_request` raises an error if the request fails. 
This means for any successful response, we should have a response result.

I updated the code in commit 77379b70697a17549d3313dd6b3d99ddcb0cc026 to be `raw_response['response']['result']`

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 3 above)
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Updated human readable output in commit 77379b70697a17549d3313dd6b3d99ddcb0cc026

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:1]: Added a short message in **bold** and regenerated the docs in commit 77379b70697a17549d3313dd6b3d99ddcb0cc026

File: Packs/PAN-OS/Integrations/Panorama/README.md
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/README.md:1]: Added a short message in **bold** to the YML and regenerated the docs in commit 77379b70697a17549d3313dd6b3d99ddcb0cc026

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 7]
```diff
@@ -14511,6 +14511,116 @@ def pan_os_list_profile_exception_command(args: dict) -> CommandResults:
...
+        )
+    else:
+        master_key_args.extend(
+            [
+                add_argument(arg=args.get('new_master_key'), field_name='new-master-key', member=False),
+                add_argument(arg=args.get('current_master_key'), field_name='current-master-key', member=False),
+            ]
+        )
+
+    master_key_args.append(add_argument_yes_no(arg='no', field_name='on-hsm'))
```
[the remark this answers is not in the document]
Comment (kamalq97) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Added a comment and improved the description in the YML / README in commit c84b49da1a093e5e8779082a770d9d5a500acecb

In the future, we should consider exposing this as an arg with two predefined values: "yes" and "no"
Currently, it is a static value, as required by the TPM.

**Comment by kamalq97** [comment]
@dorschw  There is a test playbook `palo_alto_panorama_test_pb` that intermittently fails in the nightly build (and is also failing in this branch). This issue is currently being investigated by @YaelShamai.

The update of the credentials in GSM (tied to this PR) **fixes three test playbooks** that are currently failing in the nightly due to invalid credentials.

I have added a **ForceMerge** label to this PR.

**Comment by kamalq97** [comment]
@dorschw There is a test playbook `palo_alto_panorama_test_pb` that intermittently fails in the nightly build (and is also failing in this branch). This issue is currently being investigated by @YaelShamai.

This PR fixes three test playbooks that are currently failing in the nightly build.

I have added a **ForceMerge** label to this PR.

**Pull request merged**